initrd/etc/functions

```diff
@@ -267,6 +267,19 @@ combine_configs() {
 	cat /etc/config* > /tmp/config
 }
 
+# Generate secret value using first 40 chars of ROM SHA256 hash
+secret_from_rom_hash() {
+	local ROM_IMAGE="/tmp/coreboot-notpm.rom"
+
+	echo -e "\nTPM not detected; measuring ROM directly\n" 1>&2
+	# use a previously-copied image if it exists
+	if [ -f ${ROM_IMAGE} ]; then
+		sha256sum ${ROM_IMAGE} | cut -f1 -d ' ' | cut -c 1-40 | tr -d '\n'
+	else
+		flash.sh -s ${ROM_IMAGE} | cut -c 1-40 | tr -d '\n'
+	fi
+}
```

Comment:
@MrChromebox @daringer @jans23 @kylerankin @szszszsz @osresearch @fhvyhjriur @Tonux599 @Thrilleratplay @irelativism @blobless @lrvick

1. Boot hardware secured by Librem Key/Nitrokey Pro/Nitrokey Storage on TPM-less supported devices. User owning those devices trusted marketing speech and left said hardware unattended because remotely attested with the USB Security dongle they key in their pockets. Press 'r' frenetically until reaching recovery shell. Plug formatted usb thumbdrive. Keep calm.
2. Run the following from shell prompt/a script: `mount-usb rw && flash.sh -r /media/backuped.rom && umount /media && poweroff`. Wonder around, get back your USB dongle when screen goes off. Leave the scene with USB thumbdrive.
3. Craft your own evil maid aimed rom at leasure. Tamper also with /boot content! If there is enough space in SPI flash of the hardware in question, add as much stuff as you need! make scripts to copy them under /boot if not existing, modify Heads scripts to your liking so that signed digest verification lies... If you want, you can craft an evil xen, kernel, initrd and grub config if you'd like! Be creative.
4. Obtain secret from rom at leasure, used as remote attestation secret in the HOTP challenge with user's Nitrokey Pro/Nitrokey Storage/Librem Key, by running the following on obtained backuped.rom: `sha256sum backuped.rom | cut -f1 -d ' ' | cut -c 1-40 | tr -d '\n'`
5. Replace the content of `secret_from_rom_hash` function commented here with the value obtained from precedent step, simply replacing `secret` in between the following quotes: `echo "secret"`
6. Build your crafted rom. Put it back on usb thumbdrive as file `new.rom`
7. Redo step 1 on unattended hardware because trusted by user that the rom is protected with his USB Security dongle kept securely with him.
8. Run the following from shell/script: `mount-usb rw && flash.sh /media/new.rom && umount /media && poweroff`. This will take more time then step 2, since writing takes more time. Keep looking discretely at the screen for it to go dark. Get back your thumbdrive and leave for good.
9- Be confident that on every next boot, HOTP challenge will succeed over Nitrokey Pro/Nitrokey Storage/Librem Key... until the user decides to upgrade firmware on said trusted hardware. Repeat at will on that hardware. Code was merged upstream under Heads. Keep in mind that all TPM-less hardware out there now supported by Heads can be hacked easily the same way by users bragging about their security dongles and the hardware you know doesn't have a TPM. Keep track of supported devices upstream, own them and test locally for guaranteed success!

We merge?

Reply:
hilarious (although realistic) story @tlaurion , my favorite: 
>  Play dumb. Be cute. Be prepared, plan accordingly

but seriously: seeing which misconceptions this (not even upstream) patch has already produced,
I think I also re-cast my vote to "no". I still see a gap between Heads and coreboot+seabios in terms of security (no password, no hotp, not even the chance to check the firmware, no /boot signing), but have to admit that the whole: Heads-implies-another-level-of-security argument, is stronger than I expected.